openedx/core/djangoapps/agreements/api.py

```diff
@@ -68,3 +73,180 @@ def get_integrity_signatures_for_course(course_id):
     """
     course_key = CourseKey.from_string(course_id)
     return IntegritySignature.objects.filter(course_key=course_key)
+
+
+def create_lti_pii_signature(username, course_id, lti_tools):
+    """
+    Creates an lti pii tool signature. If the signature already exist, do not create a new one.
+
+    Arguments:
+        * course_key (str)
+        * lti_tools (dict)
+        * lti_tools_hash (int)
+    Returns:
+        * An LTIPIISignature, or None if a signature already exists.
+    """
+    user = User.objects.get(username=username)
+    course_key = CourseKey.from_string(course_id)
+    lti_tools_hash = hash(str(lti_tools))
+
+    # if user and course exists, update, otherwise create a new signature
+    try:
+        LTIPIISignature.objects.get(user=user, course_key=course_key)
+    except ObjectDoesNotExist:
+        #create
+        signature = LTIPIISignature.objects.create(
+            user=user,
+            course_key=course_key,
+            lti_tools=lti_tools,
+            lti_tools_hash=lti_tools_hash)
+        return signature
+    else:
+        signature = LTIPIISignature.objects.update(
+            user=user,
+            course_key=course_key,
+            lti_tools=lti_tools,
+            lti_tools_hash=lti_tools_hash)
+        return signature
+
+
+def get_lti_pii_signature(username, course_id):
+    """
+    Get the lti pii signature of a user in a course.
+
+    Arguments:
+        * username (str)
+        * course_id (str)
+
+    Returns:
+        * An LTIPIISignature object, or None if one does not exist for the
+          user + course combination.
+    """
+    course_key = CourseKey.from_string(course_id)
+    try:
+        user = User.objects.get(username=username)
+        signature = LTIPIISignature.objects.get(user=user, course_key=course_key)
+    except ObjectDoesNotExist:
+        return None
+    else:
+        return LTIPIISignatureData(user=signature.user, course_id=str(signature.course_key),
+                                   lti_tools=signature.lti_tools, lti_tools_hash=signature.lti_tools_hash)
+
+
+def get_pii_receiving_lti_tools(course_id):
+    """
+    Get a course's LTI tools that share PII.
+
+    Arguments:
+        * course_id (str)
+
+    Returns:
+        * A List of LTI tools sharing PII.
+    """
+
+    course_key = CourseKey.from_string(course_id)
+    try:
+        course_ltipiitools = LTIPIITool.objects.get(course_key=course_key).lti_tools
+        return LTIToolsReceivingPIIData(
+            lii_tools_receiving_pii=course_ltipiitools,
+        )
+    except ObjectDoesNotExist:
+        return None
+
+
+def user_lti_pii_signature_needed(username, course_id):
+    """
+    Determines if a user needs to acknowledge the LTI PII Agreement
+
+    Arguments:
+        * username (str)
+
+    Returns:
+        * True if the user needs to sign a new acknowledgement.
+        * False if the acknowledgements are up to date.
+    """
+    if _course_has_lti_pii_tools(course_id):
```

Comment:
This looks good and captures most of the complex logic of determining whether a user needs an LTI PII signature - suggested a small correction below. Nice job!

I think we could simplify this a little bit. I don't think we need three nested if statements. We can accomplish this without that. You'll notice that there are a few duplicated return statements.

To figure out how to write this more succinctly, I created a matrix of the boolean values of `has_lti_pii_tools`, `lti_pii_signature_exists`, and `lti_pii_signature_needs_update` and how they they all resolve to the value of `user_needs_lti_pii_signature`.

That led me to write something like this...

```
course_has_lti_pii_tools = _course_has_lti_pii_tools(course_id)
user_has_signature = _user_lti_pii_signature_exists(username, course_id)
signature_needs_update = _user_needs_signature_update(username, course_id)

if course_has_lti_pii_tools && (not user_has_signature || signature_needs_update):
    return True
else:
    return False
```

What do you think of that? You could try creating your own matrix of booleans and see if this does the same thing your code does.

Reply:
I'll try out this approach!